### PR TITLE
chore: Implement State Screenshot testing utility

### LINKED ISSUE #341: Implement Matrix Screenshot Util

### PR DESCRIPTION
**Pull request by JoCa96**
Implemented a playwright utility that allwos us to create a screenshot of a matrix with all relevant component states and variations. This will make it easier for UX to review the screenshots and reduce test execution time and screenshot storage.

Added screenshot tests for #222 

closes #341

### CODE REVIEW

File: packages/sit-onyx/src/components/OnyxButton/OnyxButton.ct.tsx [hunk 1]
```diff
@@ -48,3 +49,23 @@ test("should render button with icon", async ({ mount, makeAxeBuilder }) => {
...
+
+test(
+  "Screenshot matrix",
+  createMatrixScreenshot(STATES, "matrix.png", ({ variation, state, mode }) => (
+    <OnyxButton
+      label={"label"}
+      variation={variation}
+      mode={mode}
+      disabled={state === "disabled"}
+      icon={state === "icon" ? happyIcon : ""}
```
Comment (larsrickert) [inline packages/sit-onyx/src/components/OnyxButton/OnyxButton.ct.tsx:68]:
```suggestion
      label="label"
      variation={variation}
      mode={mode}
      disabled={state === "disabled"}
      icon={state === "icon" ? happyIcon : undefined}
```

File: packages/sit-onyx/src/utils/playwright.tsx [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+ * "colHeadline1 case3 case4 case5"
+ * "colHeadline2 case6 case7 case8"
+ * "colHeadline3 case9 case10 case11"
+ * ```
+ */
+const buildGridTemplate = (rows: number, columns: number) =>
+  Array.from({ length: rows + 1 })
+    .map((_, ri) =>
+      Array.from({ length: columns + 1 })
+        .map((_, ci) => {
```
Comment (larsrickert) [inline packages/sit-onyx/src/utils/playwright.tsx:58]: Would prefer to call the variables "row" and "column" so its easier understandable what they are

File: packages/sit-onyx/src/utils/playwright.tsx [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+type CaseBuilder<S extends ComponentStates> = (c: Permutation<S>, i: number) => JSX.Element;
+
+/**
+ * Generates a test case in the grid and places it accordingly.
+ */
+const createTestCases = <S extends ComponentStates>(
+  testCases: Permutation<S>[],
+  caseBuilder: CaseBuilder<S>,
+) =>
+  testCases.map((tC, i) => (
```
Comment (larsrickert) [inline packages/sit-onyx/src/utils/playwright.tsx:79]: Would prefer "testCase" instead of "tc"

File: packages/sit-onyx/src/utils/playwright.tsx [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+    <div
+      style={{ gridArea: `case${i}` }}
+      key={i}
+      // Handle common cases
+      class={{
+        [`pw-${tC["focusState"]}`]: tC["focusState"],
```
Comment (larsrickert) [inline packages/sit-onyx/src/utils/playwright.tsx:85]: Would prefer a data attribute like `data-test-${tc.focusState}` so its clear that it is only used for testing

File: packages/sit-onyx/src/utils/playwright.tsx [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+/**
+ * Generates the jsx for the screenshot matrix test.
+ * It uses a grid with named areas to place all test cases, the row and column headlines.
+ */
+const generateScreenshotMatrix = <S extends ComponentStates>(
+  states: S,
+  caseBuilder: CaseBuilder<S>,
+) => {
+  const [rowStates] = Object.values(states);
+  const colStates = generatePermutations(Object.fromEntries(Object.entries(states).slice(1))).map(
```
Comment (larsrickert) [inline packages/sit-onyx/src/utils/playwright.tsx:102]: Imho this line is a little hard to understand, maybe a short explanatory comment or extract to a variable might help?

File: packages/sit-onyx/src/utils/playwright.tsx [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+        ))
+      }
+      {createTestCases(testCases, caseBuilder)}
+    </main>
+  );
+};
+
+type TestArg = Parameters<Parameters<typeof test>[1]>[0];
+
+export const createMatrixScreenshot =
```
Comment (larsrickert) [inline packages/sit-onyx/src/utils/playwright.tsx:157]: Some JSDoc would be nice so its easier when other devs use this in the future

**Review by larsrickert — CHANGES_REQUESTED** [review]

File: packages/sit-onyx/src/components/OnyxCheckbox/OnyxCheckbox.ct.tsx [hunk 7]
```diff
@@ -213,3 +214,23 @@ for (const testCase of invalidTestCases) {
     await expect(component).toHaveScreenshot(`invalid-${testCase.name}.png`);
   });
 }
+
+const STATES = {
+  state: ["default", "disabled", "invalid", "required", "optional"],
```
Comment (larsrickert) [inline packages/sit-onyx/src/components/OnyxCheckbox/OnyxCheckbox.ct.tsx:219]: Invalid screenshots does not seem to work, they should be read.
Invalid only triggers when the checkbox is touched so it needs to be blurred first

**Comment by larsrickert** [comment]
In general I like the matrix screenshots but I am concerned that the reliability is no longer guaranteed then...because then we  are basically only testing custom CSS selectors instead of actually hovering etc.

File: packages/sit-onyx/src/utils/playwright.tsx (hunk 6 above)
Comment (JoCa96) [reply packages/sit-onyx/src/utils/playwright.tsx:188]: done

File: packages/sit-onyx/src/utils/playwright.tsx (hunk 4 above)
Comment (JoCa96) [reply packages/sit-onyx/src/utils/playwright.tsx]: done

**Review by larsrickert — APPROVED** [review]

**Pull request merged**